### DISCUSSION TITLE
Transfer data from one page to another

### DISCUSSION THREAD

**Discussion by DoctorPok42** [vercel/next.js#28106] [Help] [answered]
Hello,
I created a site in NextJs, on which you can connect with your Discord account and I was then able (in the oauth file) to retrieve the user's guilds.

And I would like to send these guilds (json file) to my dashboard page.

oauth.tsx :
```
export default async (req: NextApiRequest, res: NextApiResponse) => {
...
  const guildsuser = await fetch("http://discord.com/api/users/@me/guilds", {
    headers: { Authorization: `${token_type} ${access_token}` },
  });
  const guilds = guildsuser.json();
...
  res.redirect("/dashboard");
}
```
and I would like to be able to export the guilds element to use it in my dashboard.

**Comment by jamesmosier** [comment, marked as answer]
There is no built-in way to send data between pages in Next. 

One option would be to save the data in localStorage. This would require that you return the data via `getStaticProps` or `getServerSideProps` to your component and then you could do something like: `window.localStorage.setItem('GUILDS', guilds)` and then retrieve it by doing `window.localStorage.getItem('GUILDS')` on your dashboard page.

Another option might be to use React's Context API to share data between pages. [This example](https://github.com/vercel/next.js/tree/canary/examples/with-context-api) shows a bit about how it is done, but in summary you could create a `_app.tsx` file with a Provider and then you could set the data in state in your provider from your `oauth` component.

**Reply by DoctorPok42** [reply]
I am ok for the getServerSideProps but I cannot put the `export async function getServerSideProps` in another export.
So how could I do ?

And then for `window.localStorage.getItem ('GUILDS')`
The console tells me that window is not defined, and that I don't know how to define it.

**Reply by harshcut** [reply]
You can't export `getServerSideProps` as default. It has be a named export. As for the `window` object, use the `useEffect` hook. The code inside `useEffect` is only executed on the client (in the browser), thus it has access to `window`.

> Next.js is universal, which means it executes code first server-side, then client-side. The window object is only present client-side, so if you absolutely need to have access to it in some React component, you should put that code in `componentDidMount`. This lifecycle method will only be executed on the client. You may also want to check if there isn't some alternative universal library which may suit your needs.

**Reply by sergioviniciuss** [reply]
@DoctorPok42 , you're getting this msg because window object is only available in the browser. On the server side, it doesn't exist. If you're calling this from a component, you can use the dynamic import from Next, passing the ssr: false flag, so this code doesn't run on server. If this is not the case, you can add a check to prevent running this code if window is not defined :)
